File: 프로그래머스/LV2/코딩테스트 연습/연습문제/3 x n 타일링/3 x n 타일링.cpp
#include <string>
#include <vector>

using namespace std;

int solution(int n) {
    long tiles[5001] = {0,};
    long mod = 1000000007;
    tiles[0] = 1;
    tiles[2] = 3;

    for(int i = 4; i <= n; i += 2)
        tiles[i] = (tiles[i-2]*4 % mod - tiles[i-4] % mod + mod) % mod;

    return tiles[n];
}

File: 프로그래머스/LV2/코딩테스트 연습/연습문제/3 x n 타일링/3 x n 타일링.py
def solution(n):
    tiles = [0] * 5001

    mod = 1000000007
    tiles[0] = 1
    tiles[2] = 3
    
    for i in range(4, n+1, 2):
        tiles[i] = (tiles[i-2]*4 % mod - tiles[i-4] % mod + mod) % mod

    return tiles[n]